File: practices/12-multi-file-refactoring/src/order-repository.ts
import { Order, OrderItem, Callback } from './types';
import * as db from './database';

const COLLECTION = 'orders';

export function initOrderRepo(callback: Callback<void>): void {
  db.initCollection(COLLECTION, callback);
}

export function createOrder(
  userId: string,
  items: OrderItem[],
  callback: Callback<Order>
): void {
  const total = items.reduce((sum, item) => sum + item.price * item.quantity, 0);

  const order: Order = {
    id: `order_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
    userId,
    items,
    total,
    status: 'pending',
    createdAt: new Date(),
  };

  db.insert(COLLECTION, order.id, order, (error, result) => {
    if (error) {
      callback(error);
      return;
    }
    callback(null, result as Order);
  });
}

export function getOrderById(
  orderId: string,
  callback: Callback<Order>
): void {
  db.findById(COLLECTION, orderId, (error, result) => {
    if (error) {
      callback(error);
      return;
    }
    callback(null, result as Order);
  });
}

export function getOrdersByUserId(
  userId: string,
  callback: Callback<Order[]>
): void {
  db.findAll(COLLECTION, (error, results) => {
    if (error) {
      callback(error);
      return;
    }
    const userOrders = (results as Order[]).filter((o) => o.userId === userId);
    callback(null, userOrders);
  });
}

export function updateOrderStatus(
  orderId: string,
  status: Order['status'],
  callback: Callback<Order>
): void {
  db.update(COLLECTION, orderId, { status }, (error, result) => {
    if (error) {
      callback(error);
      return;
    }
    callback(null, result as Order);
  });
}

File: practices/12-multi-file-refactoring/src/order_repository.py
import time
import random
import string
from typing import Any, Callable, Dict, List, Optional

import database as db

COLLECTION = "orders"

Callback = Callable[[Optional[Exception], Any], None]


def init_order_repo(callback: Callback) -> None:
    db.init_collection(COLLECTION, callback)


def create_order(
    user_id: str, items: List[Dict[str, Any]], callback: Callback
) -> None:
    total = sum(item["price"] * item["quantity"] for item in items)
    random_suffix = "".join(random.choices(string.ascii_lowercase + string.digits, k=9))

    order = {
        "id": f"order_{int(time.time() * 1000)}_{random_suffix}",
        "user_id": user_id,
        "items": items,
        "total": total,
        "status": "pending",
        "created_at": time.time(),
    }

    def _on_insert(error: Optional[Exception], result: Any) -> None:
        if error:
            callback(error, None)
            return
        callback(None, result)

    db.insert(COLLECTION, order["id"], order, _on_insert)


def get_order_by_id(order_id: str, callback: Callback) -> None:
    def _on_find(error: Optional[Exception], result: Any) -> None:
        if error:
            callback(error, None)
            return
        callback(None, result)

    db.find_by_id(COLLECTION, order_id, _on_find)


def get_orders_by_user_id(user_id: str, callback: Callback) -> None:
    def _on_find(error: Optional[Exception], results: Any) -> None:
        if error:
            callback(error, None)
            return
        user_orders = [o for o in results if o.get("user_id") == user_id]
        callback(None, user_orders)

    db.find_all(COLLECTION, _on_find)


def update_order_status(
    order_id: str, status: str, callback: Callback
) -> None:
    def _on_update(error: Optional[Exception], result: Any) -> None:
        if error:
            callback(error, None)
            return
        callback(None, result)

    db.update(COLLECTION, order_id, {"status": status}, _on_update)
